libbeat/template/template.go

```diff
@@ -184,116 +183,156 @@ func (t *Template) GetPattern() string {
+	major := esVersion.Major
 	switch {
+	case major == 2:
+		mapping.Put("_all.norms.enabled", false)
+		mapping = common.MapStr{
+			"_default_": mapping,
+		}
 	case major < 6:
-		mappingName = "_default_"
+		mapping = common.MapStr{
+			"_default_": mapping,
+		}
+	case major == 6:
+		mapping = common.MapStr{
+			"doc": mapping,
+		}
 	case major >= 7:
-		defaultFields = append(defaultFields, "fields.*")
-		indexSettings.Put("query.default_field", defaultFields)
+		// keep typeless structure
 	}
 
-	indexSettings.DeepUpdate(t.config.Settings.Index)
-
-	// Load basic structure
-	basicStructure := common.MapStr{
-		"mappings": common.MapStr{
-			mappingName: common.MapStr{
-				"_meta": common.MapStr{
-					"version": t.beatVersion.String(),
-				},
-				"date_detection":    defaultDateDetection,
-				"dynamic_templates": dynamicTemplates,
-				"properties":        properties,
-			},
-		},
-		"order": 1,
-		"settings": common.MapStr{
-			"index": indexSettings,
+	return mapping
+}
+
+func buildDynTmpl(ver common.Version) common.MapStr {
+	strMapping := common.MapStr{
+		"ignore_above": 1024,
+		"type":         "keyword",
+	}
+	if ver.Major == 2 {
+		strMapping["type"] = "string"
+		strMapping["index"] = "not_analyzed"
+	}
+
+	return common.MapStr{
+		"strings_as_keyword": common.MapStr{
+			"mapping":            strMapping,
+			"match_mapping_type": "string",
 		},
 	}
+}
 
-	if len(t.config.Settings.Source) > 0 {
-		key := fmt.Sprintf("mappings.%s._source", mappingName)
-		basicStructure.Put(key, t.config.Settings.Source)
+func buildIdxSettings(ver common.Version, userSettings common.MapStr) common.MapStr {
+	indexSettings := common.MapStr{
+		"refresh_interval": "5s",
+		"mapping": common.MapStr{
+			"total_fields": common.MapStr{
+				"limit": defaultTotalFieldsLimit,
+			},
+		},
 	}
 
-	// ES 6 moved from template to index_patterns: https://github.com/elastic/elasticsearch/pull/21009
-	if t.esVersion.Major >= 6 {
-		basicStructure.Put("index_patterns", []string{t.GetPattern()})
-	} else {
-		basicStructure.Put("template", t.GetPattern())
+	// number_of_routing shards is only supported for ES version >= 6.1
+	version61, _ := common.NewVersion("6.1.0")
+	if !ver.LessThan(version61) {
+		indexSettings.Put("number_of_routing_shards", defaultNumberOfRoutingShards)
 	}
 
-	if t.esVersion.IsMajor(2) {
-		basicStructure.Put("mappings._default_._all.norms.enabled", false)
+	if ver.Major >= 7 {
+		// copy defaultFields, as defaultFields is shared global slice.
+		fields := make([]string, len(defaultFields))
+		copy(fields, defaultFields)
+		fields = append(fields, "fields.*")
+
+		indexSettings.Put("query.default_field", fields)
 	}
 
-	return basicStructure
+	indexSettings.DeepUpdate(userSettings)
+	return indexSettings
 }
 
-func appendFields(fields, appendFields common.Fields) (common.Fields, error) {
-	if len(appendFields) > 0 {
-		appendFieldKeys := appendFields.GetKeys()
+func concatFields(a, b common.Fields) (common.Fields, error) {
```

Comment:
I think this function is general enough to be moved to `libbeat/common/fields.go`.

Reply:
moved + added cleaned up fields_test.go a little.